cpp/include/raft/core/detail/copy.hpp

```diff
@@ -0,0 +1,524 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <cstdio>
+#include <execution>
+#include <raft/core/cuda_support.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/resource/stream_view.hpp>
+#include <raft/core/resources.hpp>
+#include <type_traits>
+#ifndef RAFT_DISABLE_CUDA
+#include <raft/core/cudart_utils.hpp>
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/resource/cublas_handle.hpp>
+#include <raft/linalg/detail/cublas_wrappers.hpp>
+#ifdef __CUDACC__
+#include <raft/util/cuda_dev_essentials.cuh>
+#endif
+#endif
+
+namespace raft {
+namespace detail {
+
+template <bool B, typename DstType = void, typename SrcType = void, typename T = void>
+struct mdspan_copyable {};
+
+/*
+ * A helper struct used to determine whether one mdspan type can be copied to
+ * another and if so how
+ */
+template <typename DstType, typename SrcType, typename T>
+struct mdspan_copyable<true, DstType, SrcType, T> {
+  using dst_type = std::remove_reference_t<DstType>;
+  using src_type = std::remove_reference_t<SrcType>;
+
+  // Extents properties
+  using dst_extents_type = typename dst_type::extents_type;
+  using src_extents_type = typename src_type::extents_type;
+  using index_type =
+    std::conditional_t<(std::numeric_limits<typename dst_extents_type::index_type>::max() >
+                        std::numeric_limits<typename src_extents_type::index_type>::max()),
+                       typename dst_extents_type::index_type,
+                       typename src_extents_type::index_type>;
+
+  // Dtype properties
+  using dst_value_type                   = typename dst_type::value_type;
+  using src_value_type                   = typename src_type::value_type;
+  using dst_element_type                 = typename dst_type::element_type;
```

Comment:
Maybe another `static_assert` here that `dst_type::element_type` is non-const?

Reply:
Ah, good point! We should really use `is_assignable` rather than `is_convertible` for the compatibility check. That should take care of the const issue.